systems/robotInterfaces/@Biped/getPelvisMotionForWalking.m

```diff
@@ -0,0 +1,218 @@
+  if body_id < 0
+    body_id = obj.getFrame(body_id).body_ind;
+  end
+  if body_id == obj.foot_body_id.left
+    lfoot_body_motion = foot_motion_data(j);
+  elseif body_id == obj.foot_body_id.right
+    rfoot_body_motion = foot_motion_data(j);
+  end
+end
+
+lfoot_frame = lfoot_body_motion.eval(0);
+T_l_frame = poseQuat2tform([lfoot_frame(1:3); expmap2quat(lfoot_frame(4:6))]);
+lsole_des = tform2poseQuat(T_l_frame * T_frame_to_sole.left);
+
+rfoot_frame = rfoot_body_motion.eval(0);
+T_r_frame = poseQuat2tform([rfoot_frame(1:3); expmap2quat(rfoot_frame(4:6))]);
+rsole_des = tform2poseQuat(T_r_frame * T_frame_to_sole.right);
+
+pelvis_reference_height(1) = min(lsole_des(3),rsole_des(3));
+
+for i=1:length(support_times)-1
+  isDoubleSupport = any(supports(i).bodies==obj.foot_body_id.left) && any(supports(i).bodies==obj.foot_body_id.right);
+  isRightSupport = ~any(supports(i).bodies==obj.foot_body_id.left) && any(supports(i).bodies==obj.foot_body_id.right);
+  isLeftSupport = any(supports(i).bodies==obj.foot_body_id.left) && ~any(supports(i).bodies==obj.foot_body_id.right);
+  if options.debug
+    plot(support_times(i:i+1), 0.15 + 0.05*(isRightSupport|isDoubleSupport)*[1, 1], 'go:')
+    plot(support_times(i:i+1), 0.15 + 0.05*(isLeftSupport|isDoubleSupport)*[1,1], 'ro:')
+  end
+
+  nextIsDoubleSupport = any(supports(i+1).bodies==obj.foot_body_id.left) && any(supports(i+1).bodies==obj.foot_body_id.right);
+  nextIsRightSupport = ~any(supports(i+1).bodies==obj.foot_body_id.left) && any(supports(i+1).bodies==obj.foot_body_id.right);
+  nextIsLeftSupport = any(supports(i+1).bodies==obj.foot_body_id.left) && ~any(supports(i+1).bodies==obj.foot_body_id.right);
+
+  t = support_times(i);
+  t_next = support_times(i+1);
+
+  lfoot_frame = lfoot_body_motion.eval(t);
+  T_l_frame = poseQuat2tform([lfoot_frame(1:3); expmap2quat(lfoot_frame(4:6))]);
+  lsole_des = tform2poseQuat(T_l_frame * T_frame_to_sole.left);
+
+  rfoot_frame = rfoot_body_motion.eval(t);
+  T_r_frame = poseQuat2tform([rfoot_frame(1:3); expmap2quat(rfoot_frame(4:6))]);
+  rsole_des = tform2poseQuat(T_r_frame * T_frame_to_sole.right);
+
+  lfoot_frame_next = lfoot_body_motion.eval(t_next);
+  T_l_frame_next = poseQuat2tform([lfoot_frame_next(1:3); expmap2quat(lfoot_frame_next(4:6))]);
+  lsole_des_next = tform2poseQuat(T_l_frame_next * T_frame_to_sole.left);
+
+  rfoot_frame_next = rfoot_body_motion.eval(t_next);
+  T_r_frame_next = poseQuat2tform([rfoot_frame_next(1:3); expmap2quat(rfoot_frame_next(4:6))]);
+  rsole_des_next = tform2poseQuat(T_r_frame_next * T_frame_to_sole.right);
+
+  step_height_delta_threshold = 0.05; % cm, min change in height to classify step up/down
+  step_up_pelvis_shift = 0.07; % cm
+  if isDoubleSupport && nextIsDoubleSupport
+    pelvis_reference_height(i+1) = pelvis_reference_height(i);
+  elseif isDoubleSupport && nextIsLeftSupport
+    if lsole_des_next(3) > rsole_des(3) + step_height_delta_threshold
+      % stepping up with left foot
```

Comment:
if isDoubleSupport && nextIsLeftSupport, then wouldn't that mean that the right foot is about to move, not the left? feels like the logic is flipped here.